deps/sk_hdr_png/include/sk_hdr_png.hpp

```diff
@@ -0,0 +1,993 @@
+
+					// Multiply by 64 to get 10-bit range (0-1023) into safe 16-bit range (0-65535)
+					const uint16_t r = (((( rgba & 0x000003FF)        + 1) * 64) & 0xFFFF) - 1;
+					const uint16_t g = (((((rgba & 0x000FFC00) >> 10) + 1) * 64) & 0xFFFF) - 1;
+					const uint16_t b = (((((rgba & 0x3FF00000) >> 20) + 1) * 64) & 0xFFFF) - 1;
+
+					*png_pixels++ = r;
+					*png_pixels++ = g;
+					*png_pixels++ = b;
+
+					XMVECTOR v =
+					  XMVectorSet (static_cast <float>(r) / 65535.0f,
+					               static_cast <float>(g) / 65535.0f,
+					               static_cast <float>(b) / 65535.0f, 1.0f);
+
+					v =
+					  XMVector3Transform (
+					    PQToLinear (XMVectorSaturate (v)), c_from2020toXYZ
+					  );
+
+					*luminance++ =
+					  XMVectorGetY (v);
+				}
+
+				hr = bitmap_frame->WritePixels(height, row_stride, buffer_size, png_buffer);
+			}
+
+			else if (format == format::r16g16b16a16_float)
+			{
+				const int scrgb_postscale = 1UL << quantization_bits;
+				const float scrgb_prescale = static_cast<float>(scrgb_postscale);
+
+				uint16_t* png_pixels = (uint16_t *)png_buffer;
+				uint16_t* src_pixels = (uint16_t *)pixels;
+
+				auto luminance = luminance_array;
+
+				for (size_t y = 0; y < height; y++)
+				{
+					XMFLOAT4* rgba32_pixels = rgba32_scanline;
+
+					XMConvertHalfToFloatStream (
+					  (float *)rgba32_pixels, sizeof (float),
+					           src_pixels,    sizeof (HALF), 4 * width
+					);
+
+					for (size_t x = 0; x < width ; x++)
+					{
+						XMVECTOR rgb =
+							XMLoadFloat4 (rgba32_pixels++);
+
+						*luminance++ =
+							XMVectorGetY (XMVector3Transform (rgb, c_from709toXYZ));
+
+						rgb =
+							LinearToPQ (XMVectorMax (XMVector3Transform (rgb, c_from709to2020), g_XMZero));
+
+						if (quantization_bits != 16) {
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (std::roundf ((XMVectorGetX (rgb) * scrgb_prescale)) * 65536.0f) / scrgb_postscale));
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (std::roundf ((XMVectorGetY (rgb) * scrgb_prescale)) * 65536.0f) / scrgb_postscale));
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (std::roundf ((XMVectorGetZ (rgb) * scrgb_prescale)) * 65536.0f) / scrgb_postscale));
+						} else {
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (XMVectorGetX (rgb) * 65536.0f)));
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (XMVectorGetY (rgb) * 65536.0f)));
+							*(png_pixels++) = static_cast <uint16_t> (std::min (65535, static_cast <int> (XMVectorGetZ (rgb) * 65536.0f)));
+						}
+
+						src_pixels += 4;
+					}
+				}
+
+				hr = bitmap_frame->WritePixels(height, row_stride, buffer_size, png_buffer);
+			}
+		}
+
+		else
+		{
+			hr = E_OUTOFMEMORY;
+		}
+	}
+
+	if (SUCCEEDED(hr)) hr = bitmap_frame->Commit();
+	if (SUCCEEDED(hr)) hr = encoder->Commit();
```

Comment:
This would crash if factory creation succeeded, but the `_aligned_malloc` allocations above failed. The `hr = E_OUTOFMEMORY` code path isn't attached to the right `if` statement to prevent that. Alternatively could just move these three `if (SUCCEEDED(hr))` blocks into the `if (png_buffer != nullptr && rgba32_scanline != nullptr && luminance_array != nullptr)` block like the others before.